pkg/api/pod/util.go

```diff
@@ -48,31 +45,27 @@ const AllContainers ContainerType = (InitContainers | Containers | EphemeralCont
 // if visiting should continue.
 type ContainerVisitor func(container *api.Container, containerType ContainerType) (shouldContinue bool)
 
-// VisitContainers invokes the visitor function with a pointer to the container
-// spec of every container in the given pod spec. If visitor returns false,
+// VisitContainers invokes the visitor function with a pointer to every container
+// spec in the given pod spec with type set in mask. If visitor returns false,
 // visiting is short-circuited. VisitContainers returns true if visiting completes,
 // false if visiting was short-circuited.
-//
-// With the default mask (zero value or DefaultContainers) VisitContainers will visit all containers
-// enabled by current feature gates. If mask is non-zero, VisitContainers will unconditionally visit
-// container types specified by mask, and no feature gate checks will be performed.
 func VisitContainers(podSpec *api.PodSpec, mask ContainerType, visitor ContainerVisitor) bool {
-	if mask == DefaultContainers || (mask&InitContainers) > 0 {
+	if mask&InitContainers != 0 {
 		for i := range podSpec.InitContainers {
 			if !visitor(&podSpec.InitContainers[i], InitContainers) {
 				return false
 			}
 		}
 	}
-	if mask == DefaultContainers || (mask&Containers) > 0 {
+	if mask&Containers != 0 {
 		for i := range podSpec.Containers {
 			if !visitor(&podSpec.Containers[i], Containers) {
 				return false
 			}
 		}
 	}
-	if (mask == DefaultContainers && utilfeature.DefaultFeatureGate.Enabled(features.EphemeralContainers)) ||
-		(mask&EphemeralContainers) > 0 {
+	if utilfeature.DefaultFeatureGate.Enabled(features.EphemeralContainers) &&
+		(mask&EphemeralContainers) != 0 {
```

Comment:
this breaks the existing behavior of AllContainers by making it once again dependent on feature gate enablement

Reply:
right, but it makes more sense to have a function instead of a type DefaultContainers.

Reply:
Sorry, I misread the previous behavior. Why would we allow consumers of this function to access ephemeral containers fi they are not enabled?

Reply:
if an existing pod has ephemeral containers persisted, there are validation paths that *should* inspect those containers, even if the feature is disabled (either because it was disabled after having been enabled, or because the current process is an n-1 API server, and a version n API server allowed the data in)

Reply:
PTAL